File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Changes/HowToSubscribeToDocumentChanges.cs
using System;

using Raven.Abstractions.Data;
using Raven.Client.Changes;
using Raven.Client.Document;
using Raven.Documentation.CodeSamples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Changes
{
	public class HowToSubscribeToDocumentChanges
	{
		private interface IFoo
		{
			#region document_changes_1
			IObservableWithTask<DocumentChangeNotification> ForDocument(string docId);
			#endregion

			#region document_changes_3
			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsInCollection(string collectionName);

			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsInCollection<TEntity>();
			#endregion

			#region document_changes_6
			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsOfType(string typeName);

			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsOfType(Type type);

			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsOfType<TEntity>();
			#endregion

			#region document_changes_9
			IObservableWithTask<DocumentChangeNotification>
				ForDocumentsStartingWith(string docIdPrefix);
			#endregion

			#region document_changes_1_1
			IObservableWithTask<DocumentChangeNotification>
				ForAllDocuments();
			#endregion
		}

		public HowToSubscribeToDocumentChanges()
		{
			using (var store = new DocumentStore())
			{
				#region document_changes_2
				IDisposable subscription = store
					.Changes()
					.ForDocument("employees/1")
					.Subscribe(
						change =>
						{
							switch (change.Type)
							{
								case DocumentChangeTypes.Put:
									// do something
									break;
								case DocumentChangeTypes.Delete:
									// do something
									break;
							}
						});
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_4
				IDisposable subscription = store
					.Changes()
					.ForDocumentsInCollection<Employee>()
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_5
				string collectionName = store.Conventions.GetTypeTagName(typeof(Employee));
				IDisposable subscription = store
					.Changes()
					.ForDocumentsInCollection(collectionName)
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_7
				IDisposable subscription = store
					.Changes()
					.ForDocumentsOfType<Employee>()
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_8
				string typeName = store.Conventions.FindClrTypeName(typeof(Employee));
				IDisposable subscription = store
					.Changes()
					.ForDocumentsOfType(typeName)
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_1_0
				IDisposable subscription = store
					.Changes()
					.ForDocumentsStartingWith("employees/1") // employees/1, employees/10, employees/11, etc.
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}

			using (var store = new DocumentStore())
			{
				#region document_changes_1_2
				IDisposable subscription = store
					.Changes()
					.ForAllDocuments() // employees/1, orders/1, customers/1, etc.
					.Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
				#endregion
			}
		}
	}
}

File: Documentation/3.5/Samples/java/src/test/java/net/ravendb/clientapi/changes/HowToSubscribeToDocumentChanges.java
package net.ravendb.clientapi.changes;

import java.io.Closeable;

import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.data.DocumentChangeNotification;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.changes.IObservable;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.utils.Observers;
import net.ravendb.samples.northwind.Employee;


public class HowToSubscribeToDocumentChanges {
  @SuppressWarnings("unused")
  private interface IFoo {
    //region document_changes_1
    public IObservable<DocumentChangeNotification> forDocument(String docId);
    //endregion

    //region document_changes_3
    public IObservable<DocumentChangeNotification> forDocumentsInCollection(String collectionName);

    public IObservable<DocumentChangeNotification> forDocumentsInCollection(Class<?> clazz);
    //endregion

    //region document_changes_6
    public IObservable<DocumentChangeNotification> forDocumentsOfType(String typeName);

    public IObservable<DocumentChangeNotification> forDocumentsOfType(Class<?> clazz);
    //endregion

    //region document_changes_9
    public IObservable<DocumentChangeNotification> forDocumentsStartingWith(String docIdPrefix);
    //endregion

    //region document_changes_1_1
    public IObservable<DocumentChangeNotification> forAllDocuments();
    //endregion
  }

  @SuppressWarnings("unused")
  public HowToSubscribeToDocumentChanges() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_2
      Closeable subscription = store
        .changes()
        .forDocument("employees/1")
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            switch (change.getType()) {
              case PUT:
                // do something
                break;
              case DELETE:
                // do something
                break;
              default:
                break;
            }
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_4
      Closeable subscription = store
        .changes()
        .forDocumentsInCollection(Employee.class)
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_5
      String collectionName = store.getConventions().getTypeTagName(Employee.class);
      Closeable subscription = store
        .changes()
        .forDocumentsInCollection(collectionName)
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_7
      Closeable subscription = store
        .changes()
        .forDocumentsOfType(Employee.class)
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_8
      String typeName = store.getConventions().getFindJavaClassName().find(Employee.class);
      Closeable subscription = store
        .changes()
        .forDocumentsOfType(typeName)
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_1_0
      Closeable subscription = store
        .changes()
        .forDocumentsStartingWith("employees/1") // employees/1, employees/10, employees/11, etc.
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }

    try (IDocumentStore store = new DocumentStore()) {
      //region document_changes_1_2
      Closeable subscription = store
        .changes()
        .forAllDocuments() // employees/1, orders/1, customers/1, etc.
        .subscribe(Observers.create(new Action1<DocumentChangeNotification>() {
          @Override
          public void apply(DocumentChangeNotification change) {
            System.out.println(change.getType() + " on document " + change.getId());
          }
        }));
      //endregion
    }
  }
}
